src/ProgrammableDevice.hh

```diff
@@ -0,0 +1,34 @@
+#ifndef PROGRAMMABLEDEVICE_HH
+#define PROGRAMMABLEDEVICE_HH
+
+#include "MSXDevice.hh"
+#include "StringSetting.hh"
+#include "TclCallback.hh"
+
+namespace openmsx {
+
+class ProgrammableDevice final : public MSXDevice
+                               , private Observer<Setting>
+{
+public:
+	explicit ProgrammableDevice(const DeviceConfig& config);
+	~ProgrammableDevice();
+
+	void writeIO(word port, byte value, EmuTime::param time) override;
+	[[nodiscard]] byte readIO(word port, EmuTime::param time) override;
+
+private:
+	void update(const Setting& setting) noexcept override;
+	void reset(EmuTime::param time) override;
```

Comment:
Technically this works. But `reset()` belongs to `readIO()` and `writeIO()`. So all 3 should be either `public` (my preference) or `private`.